src/backend/distributed/planner/shard_pruning.c

```diff
@@ -0,0 +1,1267 @@
+		}
+		else
+		{
+			prunedList = pruneOneList;
+		}
+		foundRestriction = true;
+	}
+
+	/* found no valid restriction, build list of all shards */
+	if (!foundRestriction)
+	{
+		prunedList = ShardArrayToList(cacheEntry->sortedShardIntervalArray,
+									  cacheEntry->shardIntervalArrayLength);
+	}
+
+	/*
+	 * Deep copy list, so it's independent of the DistTableCacheEntry
+	 * contents.
+	 */
+	return DeepCopyShardIntervalList(prunedList);
+}
+
+
+/*
+ * ContainsFalseClause returns whether the flattened where clause list
+ * contains false as a clause.
+ */
+bool
+ContainsFalseClause(List *whereClauseList)
+{
+	bool containsFalseClause = false;
+	ListCell *clauseCell = NULL;
+
+	foreach(clauseCell, whereClauseList)
+	{
+		Node *clause = (Node *) lfirst(clauseCell);
+
+		if (IsA(clause, Const))
+		{
+			Const *constant = (Const *) clause;
+			if (constant->consttype == BOOLOID && !DatumGetBool(constant->constvalue))
+			{
+				containsFalseClause = true;
+				break;
+			}
+		}
+	}
+
+	return containsFalseClause;
+}
+
+
+/*
+ * PrunableExpressions builds a list of all prunable expressions in node,
+ * storing them in context->pruningInstances.
+ */
+static void
+PrunableExpressions(Node *node, ClauseWalkerContext *context)
+{
+	/*
+	 * Build initial list of prunable expressions.  As long as only,
+	 * implicitly or explicitly, ANDed expressions are found, this perform a
+	 * depth-first search.  When an ORed expression is found, the current
+	 * PruningInstance is added to context->pruningInstances (once for each
+	 * ORed expression), then the tree-traversal is continued without
+	 * recursing.  Once at the top-level again, we'll process all pending
+	 * expressions - that allows us to find all ANDed expressions, before
+	 * recursing into an ORed expression.
+	 */
+	PrunableExpressionsWalker(node, context);
+
+	/*
+	 * Process all pending instances.  While processing, new ones might be
+	 * added to the list, so don't use foreach().
+	 */
+	while (context->pendingInstances != NIL)
+	{
+		PendingPruningInstance *instance =
+			(PendingPruningInstance *) linitial(context->pendingInstances);
+		PruningInstance *oldPrune = instance->instance;
+		PruningInstance *newPrune = palloc0(sizeof(PruningInstance));
+
+
+		/*
+		 * We copy the partial PruningInstance, and continue adding
+		 * information by calling PrunableExpressionsWalker() on the copy,
+		 * continuing at the the node stored in
+		 * PendingPruningInstance->continueAt.
+		 *
+		 * Check the places in PruningInstanceWalker that push onto
+		 * context->pendingInstances why construction of the PruningInstance
+		 * was deferred.
+		 *
+		 * To make the new PruningInstance useful for pruning, we have to
+		 * unset it being partial - if necessary it'll be marked so again by
+		 * PrunableExpressionsWalker().
+		 */
+		memcpy(newPrune, oldPrune, sizeof(PruningInstance));
+		newPrune->addedToPruningInstances = false;
+		newPrune->isPartial = false;
```

Comment:
For readability, it might makes sense to move the 4 lines above to something like an `InitializeWithSameConstraints` function.